Autograd Function Fallback bug fix - moe support

pengwa opened the pull request:
**Description**: Autograd Function Fallback bug fix - moe support

- Support forward inputs orders like "Non_tensor/Tensor/Non_tensor". Correspondingly, support "None/Tensor_Grad/None" for backward outputs.
- Report RuntimeError when PythonOp detected but _enable_custom_autograd_function is NOT enabled.
- Simplify the attributed used by PythonOpGrad. Renaming some attribute to reflect whether it is used for tensor inputs or all (tensor + non-tensor) inputs.

Attached the PythonOP/PythonOpGrad schemas (and how those attributes used) after the change:

![image](https://user-images.githubusercontent.com/10530022/124343983-6192fb80-dc02-11eb-9887-cb7b75335aec.png)


**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/__init__.py:
```diff
@@ -41,7 +41,7 @@
     print("WARNING: ORTModule detected PyTorch's CPP extension lock file during initialization, "
           "which can cause the script to stop responding. "
-          f"Delete {os.path.join(TORCH_CPP_BUILD_DIR,'lock')} if a hang occurs.")
+          f"Delete {os.path.join(TORCH_CPP_BUILD_DIR,'lock')} if the script stop responding and not exiting.")
```
This us unrelated to this PR and it is being completely reworked at https://github.com/microsoft/onnxruntime/pull/8027/files. Try not changing this code to prevent conflicts between both PRs

pengwa replied:
Sure, thanks for the heads up.

pengwa replied:
Fixed, please check.

wschin commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -3123,38 +3117,12 @@ Return true if all elements are true and false otherwise.
-        // The first input is a pointer which points to
-        // a Python object created by torch.autograd.Function.apply.
-        // For details, see how we interpret it in PythonOpGrad implementation.
-        for (auto i = 1; i < input_count; ++i) {
```
This doesn't look robust. I prefer to keep as many type checks as possible.

pengwa replied:
I think it is fine, because this is the shape inference of PythonOpGrad, all those attributes for example input_tensor_requires_grads comes from PythonOp's attribute. When we create PythonOpGrad, we already do the check. Does that make sense?

pengwa replied:
As we talked, I keep those attributes as redundancy, for more checks. Add some explanations in the op definitions in case of confusion of anyone curious why they are needed. :)

wschin commented on onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:
```diff
@@ -171,6 +171,12 @@ void InvokeRunner(
   // first element.
   for (; i < static_cast<size_t>(PyTuple_Size(result_ptr.get())); ++i) {
     PyObject* dl_tensor_pointer = PyTuple_GetItem(result_ptr.get(), i);
+    if (dl_tensor_pointer == Py_None) {
```
We previously filter out all `None`s from Python side. Does this change means we don't filter out `None`s anymore?

pengwa replied:
Yes, we don't filter them out, because None can be in any places of the output lists. We need map them correctly into PythonOp kernels's outputs.

tlh20 commented on onnxruntime/core/language_interop_ops/torch/torch_proxy.cc:
```diff
@@ -171,6 +171,12 @@ void InvokeRunner(
+      continue;
+    }
+
     ORT_ENFORCE(Py_REFCNT(dl_tensor_pointer) == 1, "Ref count of dl_tensor_pointer should be 1.");
```
Can the user's code cause a > 1 ref count here?

pengwa replied:
This is a good question. The dl_tensor is created by orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (when we call to_dplack, which finally PyCapsule_New a PyObject*), so it should be only affected by our code there. User code should no increase its refcnt, but can increase the forward output torch tensor

tlh20 commented on orttraining/orttraining/core/graph/gradient_builder.cc:
```diff
@@ -1805,9 +1772,24 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
     input_args.push_back(O(i));
   }
 
+  // src_attrs["input_requires_grads"] stores all inputs's requires_grad attributes,
+  // including both tensor inputs and non-tensor inputs (e.g. constants), here we filter out
+  // those non-tensor inputs when constructing PythonOpGrad's outputs.
```
input_requires_grads is generated during export, right?  If so, do we have a mechanism to detect if the requirements change during training -- say if, for some reason, the graph is initially exported with more/fewer tensors requiring gradients than?  [ Not a use case I have seen, but looking to catch this kind of issue and report and error if it does occur ]

pengwa replied:
I understand the question this way (correct me if I am wrong): we need check the tensor + non-tensor inputs counts , for all attributes like "input_requires_grads , input_convention". For this, we have some checks when PythonOp is constructed. And I just added more checks enhance it.

tlh20 commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -2887,9 +2887,14 @@ Return true if all elements are true and false otherwise.
           AttributeProto::STRING)
+      .Attr(
+          "input_requires_grads",
+          "Flags to indicate whether the torch.autograd.apply's inputs (including both tensor"
+          " and non-tensor inputs) has gradient",
```
```suggestion
          "Flags to indicate whether the torch.autograd.apply's inputs require gradients (including flags for both tensor"
          " and non-tensor inputs)",
```

pengwa replied:
Thanks alot for the formatting, my clang-format did not work at some point, let me format them in batch. :)

tlh20 commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -3063,8 +3064,10 @@ Return true if all elements are true and false otherwise.
-          "Inputs of autograd.Function.backward. There are 2*N inputs: \
-          N gradient inputs + N forward run activations of PythonOp (not including the context).",
+          "There are 2*N inputs: "
+          "  N gradient inputs(as inputs of autograd.Function.backward) + "
```
```suggestion
          "  N gradient inputs (as inputs of autograd.Function.backward) + "
```

tlh20 commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -3086,34 +3089,25 @@ Return true if all elements are true and false otherwise.
       .Attr(
           "output_tensor_ranks",
-          "Output ranks of autograd.Function.backward.",
+          "Output ranks of autograd.Function.backward outputs(including only tensor outputs).",
```
```suggestion
          "Output ranks of autograd.Function.backward outputs (including only tensor outputs).",
```

tlh20 commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -3086,34 +3089,25 @@ Return true if all elements are true and false otherwise.
       .Attr(
           "output_tensor_requires_grads",
-          "Flags to indicate which outputs has gradient",
+          "Flags to indicate which outputs(including only tensor outputs) has gradient.",
```
```suggestion
          "Flags to indicate which outputs have gradients (including only tensor outputs).",
```

tlh20 commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -3123,38 +3117,12 @@ Return true if all elements are true and false otherwise.
-        ORT_ENFORCE(output_tensor_types_proto, "PythonOp's must have \"output_tensor_types\" attribute.");
+        ORT_ENFORCE(output_tensor_types_proto, "PythonOpGrad's must have \"output_tensor_types\" attribute.");
         // Set inferred output types.
         static size_t rank_count = 0;
```
Is "static" correct here (and above, for PythonOp)?

tlh20 commented on orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:
```diff
@@ -30,11 +30,11 @@ void PythonOpBase::Init(const OpKernelInfo& info) {
   is_training_mode_ = static_cast<bool>(info.GetAttrOrDefault("training_mode", static_cast<int64_t>(0)));
-  ORT_THROW_IF_ERROR(info.GetAttr("call_convention", &call_convention_));
+  ORT_THROW_IF_ERROR(info.GetAttr("input_convention", &input_convention));
+  input_requires_grads_ = info.GetAttrsOrDefault("input_requires_grads", std::vector<int64_t>());
```
Should input_requires_grads always be defined?  If so we should omit the default and fail if it is missing (avoiding an extra allocation).

pengwa replied:
Good suggestion! input_requires_grads  should always be defined, we should fail otherwise.

pengwa replied:
Fixed all

tlh20 commented on orttraining/orttraining/training_ops/cpu/torch/torch_custom_function_kernel_base.cc:
```diff
@@ -224,11 +224,14 @@ void PythonOpBase::SetOtherOutputs(OpKernelContext* context, std::vector<OrtValu
+  ORT_ENFORCE(output_tensor_types_.size() == output_tensor_requires_grads_.size(), "backward tensor output count mismatch");
+
+  // Exclude context and forward activations from inputs
+  const size_t tensor_input_count = (info.node().InputDefs().size() - 1) / 2;
```
This looks a bit fragile if we add extra inputs at some point in the future.  We could check that InputDefs().size() is odd as one safeguard -- are there other details we can assert?

SherlockNoMad replied:
+1. Please mark a todo here. 
Also, we are including all the outputs as a work around for the seg fault issue. 
After fix, the number of input should be just N+1 .

pengwa replied:
Thanks for the feedbacks, I would admit this is really not a good one.  Combining with Wei-Sheng's comment, I think a better way is, we add some redundant attributes like "input_tensor_types_" back to easily infer the tensor input count. That would be safe once we added more inputs. 

Besides, there are other pieces similar with this. I will try to eliminate this kind of logic once getting a better way to handle the lifecycle of ctx along with forward torch tensors.

pengwa replied:
Added a todo on my head

tlh20 reviewed (changes requested):
A few minor comments on places we could extend assertion checks.  Adjacent to some of the changes in this PR there is a use of "static" that looks suspicious -- could you check if that is correct?

SherlockNoMad commented on orttraining/orttraining/core/graph/gradient_builder.cc:
```diff
@@ -1781,22 +1752,18 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
+      input_args.push_back(ArgDef());
     }
   }
 
```
Stashing all the outputs for backward computation could contribute to higher peak memory... 
I wonder if the seg fault is related to the ctx.save_for_backward(output_tensor)? 

This should be marked as an to-investigate issue.

pengwa replied:
The seg fault happens when ctx is teared down before PythonOpGrad runs. To make sure ctx is not destroyed, at least one of those forward outputs remains before PythonOpGrad kernels completes. So you are right, we should not keep all outputs if we could identify which part of them contributes to backward computations. I will follow up on it.

pengwa replied:
Added a TODO

pengwa commented on orttraining/orttraining/core/graph/gradient_builder.cc:
```diff
@@ -1805,9 +1772,24 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
+  const auto& fw_input_requires_grads = src_attrs["input_requires_grads"].ints();
+  std::vector<int64_t> bw_tensor_output_requires_grads;
+  for (auto i = 0; i < fw_input_requires_grads.size(); ++i) {
+    if (input_convention[i] == 'd') { // only handle gradients for tensor type inputs.
```
Yes we did the check here. https://github.com/microsoft/onnxruntime/blob/7ed72bcdbff2033f913bba4126ffae082367d010/orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py#L90 

If the value is not 'c' or 'd', we will fail it.

pengwa commented on orttraining/orttraining/core/graph/training_op_defs.cc:
```diff
@@ -3086,34 +3089,25 @@ Return true if all elements are true and false otherwise.
       .Attr(
           "output_tensor_types",
-          "Output types of autograd.Function.backward.",
+          "Output types of autograd.Function.backward outputs(including only tensor outputs).",
```
my clang-format did not work at some point, I will check all format in batch. Thanks a lot!

pengwa commented:
> A few minor comments on places we could extend assertion checks. Adjacent to some of the changes in this PR there is a use of "static" that looks suspicious -- could you check if that is correct?

Thanks @tlh20  for your time reviewing , I have addressed all of comments . :)
